mdagent/tools/base_tools/analysis_tools/pca_tools.py

```diff
@@ -0,0 +1,370 @@
+        self.path_registry.map_path(plot_id, path + file_name, description=desc)
+        return f"PCA plots saved as {plot_id} "
+
+    def measure_cosine_convergence(self):
+        if not self.pc:
+            pc_mess = self.get_pc()
+        else:
+            pc_mess = ""
+        if not self.n_pcs:
+            self._get_number_pcs()
+            self._make_transformation()
+
+        pc_messages = []
+        for i in range(self.n_pcs):
+            cc = self._cosine_content(self.transformed, i)
+            pc_messages.append(f"Cosine Content for PC {i+1}-{cc:.3f}\n")
+        cc_message = f"Cosine Content of each PC: {','.join(pc_messages)}"
+        return pc_mess + cc_message
+
+    def run_all(self):
+        self.get_pc()
+        try:
+            scree_plot = self.make_scree_plot()
+        except Exception as e:
+            raise Exception(f"Error during Scree Plot: str({e})")
+        try:
+            pc_plots = self.make_pc_plots()
+        except Exception as e:
+            raise Exception(f"Error during pc plots: str({e})")
+        try:
+            cos_conv = self.measure_cosine_convergence()
+        except Exception as e:
+            raise Exception(f"Error during cosine convergence: str({e})")
+        return f"Analyses done: {scree_plot}, {pc_plots}, {cos_conv}"
+
+
+class PCASchema(BaseModel):
+    trajectory_fileid: str = Field(
+        description="Trajectory File ID of the simulation to be analyzed"
+    )
+    topology_fileid: str = Field(
+        description=("Topology File ID of " "the simulation to be analyzed")
+    )
+    pc_percentage: Optional[float] = Field(
+        95.0, description="Max cumulative percentage of components for analysis"
+    )
+    analysis: str = Field(
+        "all",
+        description=(
+            "Type of analysis to be done. Availables are: "
+            "scree_plot (Saves a scree plot of the eigenvalues), "
+            "pca_analysis (gets principal components and saves"
+            " a grid plot for visualization) "
+            "Cosine (measures the cosine convergence of the top 3 PCs)"
+            "all (makes all of the previous analysis)"
+        ),
+    )
+    selection: Optional[str] = Field(
+        "backbone and name CA",
+        description=(
+            "Which selection of atoms from the simulation "
+            "to use for the pca analysis"
+        ),
+    )
+    remove_terminals: Optional[str] = Field(
+        False, description="To remove or not the terminal residues of each chain."
+    )
+
+
+class PCATool(BaseTool):
+    name = "PCATool"
+    description = (
+        "Calculate the Principal Analysis Components of a MD trajectory and "
+        "performs analysis with them"
+    )
+    args_schema = PCASchema
+    path_registry: Optional[PathRegistry]
+
+    def __init__(self, path_registry: Optional[PathRegistry] = None):
+        super().__init__()
+        self.path_registry = path_registry
+
+    def _run(self, input):
```

Comment:
can we separate this out into multiple functions? it might be easier to debug & is probably cleaner to split it up

Reply:
done! split into get_values, and run_pca_analysis functions. it reads better now